mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h

```diff
@@ -7553,6 +7577,310 @@ struct FillRewritePattern : public OpRewritePattern<miopen::FillOp> {
+
+  enum RotationDirection {
+    // left rotation by 1 is a b c d -> b c d a
+    // right rotation by 1 is a b c d -> d a b c
+    Left,
+    Right
+  };
+
+  Value emitRotations(Location loc, PatternRewriter &b, Value vector,
+                      Value laneId, RotationDirection dir, uint32_t groupSize,
+                      uint32_t totalSize,
+                      Optional<ArrayRef<uint32_t>> lanePerm) const {
+    assert(totalSize % groupSize == 0 &&
+           "block size is divisible by group size");
+
+    uint32_t logGroupSize = llvm::Log2_32_Ceil(groupSize);
+
+    int32_t base = 0, offset = 0, target = 0;
+    switch (dir) {
+    case Left:
+      base = 0;
+      offset = 1;
+      target = logGroupSize;
+      break;
+    case Right:
+      base = logGroupSize - 1;
+      offset = -1;
+      target = -1;
+      break;
+    }
+
+    Value zeroConst = b.create<ConstantIndexOp>(loc, 0);
+
+    llvm::SmallVector<Value, 4> indexConsts;
+    for (uint32_t i = 0; i < totalSize; ++i) {
+      indexConsts.push_back(b.create<ConstantIntOp>(loc, i, b.getI32Type()));
+    }
+
+    Value laneInSwizzleGroup;
+    if (lanePerm.hasValue()) {
+      Value groupSizeConst = b.create<ConstantIndexOp>(loc, 4);
+      laneInSwizzleGroup =
+          b.create<UnsignedRemIOp>(loc, laneId, groupSizeConst);
+    }
+
+    Value result = vector;
+
+    for (int32_t logRotation = base; logRotation != target;
+         logRotation += offset) {
+      uint32_t rotation = 1 << logRotation;
+      Value shouldParticipate;
+      if (lanePerm.hasValue()) {
+        // Non-standard arrangement of rows -> lanes, use longer test
+        ArrayRef<uint32_t> theLanePerm = lanePerm.getValue();
+        llvm::SmallVector<Value, 4> comparisons;
+        for (uint32_t i = 0; i < theLanePerm.size(); ++i) {
+          if ((theLanePerm[i] & rotation) != 0) {
+            Value toTest = b.create<ConstantIndexOp>(loc, i);
+            comparisons.push_back(b.create<CmpIOp>(loc, CmpIPredicate::eq,
+                                                   laneInSwizzleGroup, toTest));
+          }
+        }
+        if (comparisons.empty()) {
+          llvm_unreachable("Permutation on [0, 2^k) didn't have any entries "
+                           "with some bit set");
+        }
+        shouldParticipate =
+            std::accumulate(comparisons.begin() + 1, comparisons.end(),
+                            comparisons[0], [&b, &loc](Value v1, Value v2) {
+                              return b.create<OrOp>(loc, v1, v2);
+                            });
+      } else { // The usual case
+        Value maskConst = b.create<ConstantIndexOp>(loc, rotation);
+        Value shouldParticipateVal = b.create<AndOp>(loc, laneId, maskConst);
+        shouldParticipate = b.create<CmpIOp>(loc, CmpIPredicate::ne,
+                                             shouldParticipateVal, zeroConst);
+      }
+
+#if 0 // Compiler is busted, explicitly emit selects
```

Comment:
can we also remove this `#if 0` block?

Reply:
That one I'd like to keep in, since it's got the code I'd _like_ to use once the compiler team fixes the backend.